Reword glossary entry for cluster (#48686)
kubernetes/website: content/en/docs/reference/glossary/cluster.md

date: 2019-06-15
full_link: 
short_description: >
   A set of machines hosting containerized workloads, in the form of pods, and the control plane that manages them.  
aka: 
tags:
- fundamental
- operation
---

> A set of machines hosting containerized {{< glossary_tooltip text="workloads" term_id="workload" >}}, in the form of {{< glossary_tooltip text="pods" term_id="pod" >}}, and the {{< glossary_tooltip text="control plane" term_id="control-plane" >}} that manages them.  

Reviewer: I think we can do better. Try this:
```suggestion
A {{< glossary_tooltip text="control plane" term_id="control-plane" >}} and a set of worker machines that the
control plane manages, with the aim of running containerized {{< glossary_tooltip text="workloads" term_id="workload" >}}
as {{< glossary_tooltip text="pods" term_id="pod" >}}.
```

Author: This depends on the other comment...

Author: Line 14 seems to be missing something?

Reviewer: I revised. BTW you can accept suggestions as is, but you can also incorporate the spirit of the feedback and apply your own edits.

<!--more-->

> Machines may be physical or virtual.  Machines that host pods are referred to as {{< glossary_tooltip text="nodes" term_id="node" >}}.  Nodes that host workloads are referred to as worker nodes. There must be at least one worker node within a cluster. 

Reviewer: No, this isn't correct. A cluster with zero worker nodes isn't very useful, but it's still technically a cluster (maybe you are auto scaling).

Author: Ah - OK.  I kept this as it was part of the previous definition :)

> Control plane components run on one or more of the machines within the cluster. They may execute directly in the context of the operating system of a machine, or within pods.  When control plane components run within pods on a machine, the machine may also be referred to as a control plane node.  

Reviewer: No, this isn't correct. You can run the control plane within the cluster, but we actually recommend not doing that for the very best resilience.

Reviewer: Control plane:
- always part of the cluster
- optional to run it on nodes
- if you run it on a separate server, the thing you run on a separate server is (if healthy) still logically a control plane, and that logical thing is part of the cluster

Author: So I'm having trouble with the distinction between "part of the cluster" and "within the cluster". :)

Author: Got it.  I could co-host my control plane instance, with a bunch of other control planes for other clusters, on the same machines.  I could host my control plane within another cluster that has it's own control plane.... etc.  In tuo capite fiat!   Will nudge this and other PRs to reflect.  Thx.

Reviewer: Please do update this when you have time to.

> In production environments, the cluster typically consists of multiple machines, with the control plane components and workloads distributed across them for scale, resilience and high availability.  It is a best practice in production environments to have workloads hosted on machines that do not host control plane components.

Reviewer: We don't formally recommend this topology; we think that a control plane that runs separately from the cluster is an even more resilient choice. However it is up to the cluster operator to make that call, not this project.

>  It is a best practice in production environments to have workloads hosted on machines that do not host control plane components.

True, but doesn't belong in the glossary.  Instead, hyperlink to the page that makes the recommendation.
